File: sdks/node/src/semantic-dependency.js
import { createHash } from "node:crypto";

import {
  currentOperationContext,
  markOperationUnowned,
  openObservation,
} from "./engine-operation.js";
import { parseStrictJson } from "./managed-json.js";
import {
  NATIVE_ENGINE_MAX_OBSERVATION_CHUNK_BYTES,
  NATIVE_ENGINE_MAX_OBSERVATION_RESPONSE_READ_BYTES,
} from "./native-engine.js";

const REQUEST_FORMAT = "reproit.semantic-dependency-request.v1";
const RESPONSE_FORMAT = "reproit.semantic-dependency-response.v1";
const MAX_RECORD_BYTES = 65_536;
const MAX_TARGET_BYTES = 8 * 1_024;
const MAX_PAYLOAD_BYTES = 24 * 1_024;
const MAX_METADATA_ENTRIES = 64;
const MAX_METADATA_BYTES = 8 * 1_024;
const MAX_METADATA_NAME_BYTES = 256;
const MAX_METADATA_VALUE_BYTES = 4 * 1_024;
const COMPONENT = /^[a-z][a-z0-9.-]*$/u;
const METHOD = /^[A-Z][A-Z0-9-]{0,31}$/u;
const DIGEST = /^sha256:[0-9a-f]{64}$/u;
const ERROR_CODES = new Set([
  "interrupted",
  "invalid-input",
  "not-found",
  "other",
  "permission-denied",
  "resource-limit",
  "unsupported",
]);
const CLASS_OPERATIONS = Object.freeze({
  database: new Set(["database-execute"]),
  "outbound-http": new Set(["outbound-http-request"]),
  queue: new Set([
    "queue-acknowledge",
    "queue-publish",
    "queue-receive",
    "queue-reject",
  ]),
});
const REQUEST_KEYS = [
  "encoding",
  "format",
  "metadata",
  "method",
  "observation_class",
  "operation",
  "payload",
  "protocol",
  "target",
];
const RESPONSE_KEYS = [
  "error_code",
  "error_number",
  "format",
  "metadata",
  "observation_class",
  "operation",
  "outcome",
  "payload",
  "request_digest",
  "status",
  "status_code",
];

export function dependencyRequest(fields) {
  const request = {
    encoding: fields?.encoding,
    metadata: cloneMetadata(fields?.metadata),
    method: fields?.method,
    observationClass: fields?.observationClass,
    operation: fields?.operation,
    payload: cloneBuffer(fields?.payload),
    protocol: fields?.protocol,
    target: fields?.target,
  };
  validateRequest(request);
  return Object.freeze(request);
}

export function dependencyResponse(fields) {
  const response = {
    errorCode: fields?.errorCode ?? null,
    errorNumber: fields?.errorNumber ?? null,
    metadata: cloneMetadata(fields?.metadata),
    outcome: fields?.outcome,
    payload: fields?.payload === null ? null : cloneBuffer(fields?.payload),
    status: fields?.status ?? null,
    statusCode: fields?.statusCode ?? null,
  };
  return Object.freeze(response);
}

export function encodeDependencyRequest(request) {
  validateRequest(request);
  return boundedRecord({
    encoding: request.encoding,
    format: REQUEST_FORMAT,
    metadata: encodeMetadata(request.metadata),
    method: request.method,
    observation_class: request.observationClass,
    operation: request.operation,
    payload: request.payload.toString("base64url"),
    protocol: request.protocol,
    target: Buffer.from(request.target, "utf8").toString("base64url"),
  });
}

export function decodeDependencyRequest(record) {
  const value = parseRecord(record, REQUEST_KEYS);
  require(value.format === REQUEST_FORMAT);
  const request = dependencyRequest({
    encoding: value.encoding,
    metadata: decodeMetadata(value.metadata),
    method: value.method,
    observationClass: value.observation_class,
    operation: value.operation,
    payload: decodeBase64url(value.payload),
    protocol: value.protocol,
    target: decodeUtf8(value.target),
  });
  return request;
}

export function encodeDependencyResponse(requestRecord, response) {
  const request = decodeDependencyRequest(requestRecord);
  validateResponse(request, response);
  return boundedRecord({
    error_code: response.errorCode,
    error_number: response.errorNumber,
    format: RESPONSE_FORMAT,
    metadata: encodeMetadata(response.metadata),
    observation_class: request.observationClass,
    operation: request.operation,
    outcome: response.outcome,
    payload: response.payload === null
      ? null
      : response.payload.toString("base64url"),
    request_digest: digest(requestRecord),
    status: response.status,
    status_code: response.statusCode,
  });
}

export function decodeDependencyResponse(requestRecord, responseRecord) {
  const request = decodeDependencyRequest(requestRecord);
  const value = parseRecord(responseRecord, RESPONSE_KEYS);
  require(
    value.format === RESPONSE_FORMAT &&
    value.observation_class === request.observationClass &&
    value.operation === request.operation &&
    typeof value.request_digest === "string" &&
    DIGEST.test(value.request_digest) &&
    value.request_digest === digest(requestRecord),
  );
  const response = dependencyResponse({
    errorCode: value.error_code,
    errorNumber: value.error_number,
    metadata: decodeMetadata(value.metadata),
    outcome: value.outcome,
    payload: value.payload === null ? null : decodeBase64url(value.payload),
    status: value.status,
    statusCode: value.status_code,
  });
  validateResponse(request, response);
  return response;
}

export function runDependency(request, capture) {
  const context = currentOperationContext();
  if (context === null) return capture();
  let requestRecord;
  try {
    requestRecord = encodeDependencyRequest(request);
  } catch {
    markInvalidRequest(context, request);
    return capture();
  }
  let session;
  try {
    session = openObservation(context, request.observationClass);
  } catch {
    context.abandon();
    return capture();
  }
  if (session === null) {
    return capture();
  }
  let written;
  try {
    written = writeChunks(session, "request", requestRecord);
  } catch {
    session.abandon();
    return capture();
  }
  if (!written) return capture();
  let action;
  try {
    action = session.dispatch();
  } catch {
    session.abandon();
    return capture();
  }
  if (action === "replay") return replayDependency(session, requestRecord);
  if (action !== "capture") {
    session.abandon();
    return capture();
  }
  let captured;
  try {
    captured = capture();
  } catch (error) {
    session.abandon();
    throw error;
  }
  if (captured !== null && typeof captured?.then === "function") {
    return Promise.resolve(captured).then(
      (response) => finishCapture(session, requestRecord, response),
      (error) => {
        session.abandon();
        throw error;
      },
    );
  }
  return finishCapture(session, requestRecord, captured);
}

function markInvalidRequest(context, request) {
  if (CLASS_OPERATIONS[request?.observationClass] !== undefined) {
    markOperationUnowned(
      context,
      request.observationClass,
      Buffer.from("semantic-dependency-request-invalid", "utf8"),
    );
  } else {
    context.abandon();
  }
}

function finishCapture(session, requestRecord, response) {
  let responseRecord;
  try {
    responseRecord = encodeDependencyResponse(requestRecord, response);
  } catch {
    session.abandon();
    return response;
  }
  if (writeChunks(session, "response", responseRecord)) {
    session.finish(response.outcome);
  }
  return response;
}

function replayDependency(session, requestRecord) {
  try {
    const responseRecord = readResponse(session);
    const response = decodeDependencyResponse(requestRecord, responseRecord);
    if (!session.finish(response.outcome)) throw invalidReplay();
    return response;
  } catch (error) {
    session.abandon();
    if (error?.code === "ERR_REPROIT_SEMANTIC_DEPENDENCY") throw error;
    throw invalidReplay();
  }
}

function validateRequest(request) {
  require(request !== null && typeof request === "object");
  require(
    CLASS_OPERATIONS[request.observationClass]?.has(request.operation) === true &&
    validComponent(request.protocol, 64) &&
    validComponent(request.encoding, 64) &&
    validUtf8(request.target, MAX_TARGET_BYTES) &&
    Buffer.isBuffer(request.payload) &&
    request.payload.length <= MAX_PAYLOAD_BYTES,
  );
  validateMetadata(request.metadata);
  if (request.observationClass === "outbound-http") {
    require(typeof request.method === "string" && METHOD.test(request.method));
  } else {
    require(request.method === null);
  }
}

function validateResponse(request, response) {
  require(response !== null && typeof response === "object");
  if (response.outcome === "error") {
    require(
      ERROR_CODES.has(response.errorCode) &&
      validErrorNumber(response.errorNumber) &&
      response.payload === null &&
      Array.isArray(response.metadata) &&
      response.metadata.length === 0 &&
      response.status === null &&
      response.statusCode === null,
    );
    return;
  }
  require(
    response.outcome === "response" &&
    response.errorCode === null &&
    response.errorNumber === null &&
    Buffer.isBuffer(response.payload) &&
    response.payload.length <= MAX_PAYLOAD_BYTES,
  );
  validateMetadata(response.metadata);
  if (request.observationClass === "outbound-http") {
    require(
      response.status === null &&
      Number.isInteger(response.statusCode) &&
      response.statusCode >= 100 &&
      response.statusCode <= 599,
    );
  } else {
    require(
      response.statusCode === null &&
      (response.status === null || validComponent(response.status, 64)),
    );
  }
}

function validateMetadata(metadata) {
  require(Array.isArray(metadata) && metadata.length <= MAX_METADATA_ENTRIES);
  let total = 0;
  for (const field of metadata) {
    require(
      field !== null &&
      typeof field === "object" &&
      sameKeys(field, ["name", "value"]) &&
      validUtf8(field.name, MAX_METADATA_NAME_BYTES) &&
      Buffer.isBuffer(field.value) &&
      field.value.length <= MAX_METADATA_VALUE_BYTES,
    );
    total += Buffer.byteLength(field.name, "utf8") + field.value.length;
    require(total <= MAX_METADATA_BYTES);
  }
}

function cloneMetadata(metadata) {
  if (!Array.isArray(metadata)) return metadata;
  return metadata.map((field) => ({
    name: field?.name,
    value: cloneBuffer(field?.value),
  }));
}

function encodeMetadata(metadata) {
  return metadata.map((field) => ({
    name: Buffer.from(field.name, "utf8").toString("base64url"),
    value: field.value.toString("base64url"),
  }));
}

function decodeMetadata(value) {
  require(Array.isArray(value));
  const metadata = value.map((field) => {
    require(
      field !== null &&
      typeof field === "object" &&
      !Array.isArray(field) &&
      sameKeys(field, ["name", "value"]),
    );
    return {
      name: decodeUtf8(field.name),
      value: decodeBase64url(field.value),
    };
  });
  validateMetadata(metadata);
  return metadata;
}

function parseRecord(record, keys) {
  require(Buffer.isBuffer(record) && record.length <= MAX_RECORD_BYTES);
  let value;
  try {
    value = parseStrictJson(record, MAX_RECORD_BYTES);
  } catch {
    throw invalidRecord();
  }
  require(
    value !== null &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    sameKeys(value, keys) &&
    Buffer.from(JSON.stringify(value), "utf8").equals(record),
  );
  return value;
}

function boundedRecord(value) {
  let record;
  try {
    record = Buffer.from(JSON.stringify(value), "utf8");
  } catch {
    throw invalidRecord();
  }
  require(record.length <= MAX_RECORD_BYTES);
  return record;
}

function decodeBase64url(value) {
  require(typeof value === "string" && /^[A-Za-z0-9_-]*$/u.test(value));
  let bytes;
  try {
    bytes = Buffer.from(value, "base64url");
  } catch {
    throw invalidRecord();
  }
  require(bytes.toString("base64url") === value);
  return bytes;
}

function decodeUtf8(value) {
  const bytes = decodeBase64url(value);
  let decoded;
  try {
    decoded = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    throw invalidRecord();
  }
  return decoded;
}

function validUtf8(value, maximumBytes) {
  if (typeof value !== "string") return false;
  const bytes = Buffer.from(value, "utf8");
  return bytes.length > 0 &&
    bytes.length <= maximumBytes &&
    bytes.toString("utf8") === value;
}

function validComponent(value, maximumBytes) {
  return typeof value === "string" &&
    value.length <= maximumBytes &&
    COMPONENT.test(value);
}

function validErrorNumber(value) {
  return value === null || (
    Number.isInteger(value) && value >= 0 && value <= 0xffff_ffff
  );
}

function cloneBuffer(value) {
  return Buffer.isBuffer(value) ? Buffer.from(value) : value;
}

function digest(value) {
  return `sha256:${createHash("sha256").update(value).digest("hex")}`;
}

function writeChunks(session, stream, record) {
  for (
    let offset = 0;
    offset < record.length;
    offset += NATIVE_ENGINE_MAX_OBSERVATION_CHUNK_BYTES
  ) {
    const chunk = record.subarray(
      offset,
      Math.min(offset + NATIVE_ENGINE_MAX_OBSERVATION_CHUNK_BYTES, record.length),
    );
    const written = stream === "request"
      ? session.writeRequest(chunk)
      : session.writeResponse(chunk);
    if (!written) return false;
  }
  return true;
}

function readResponse(session) {
  const chunks = [];
  let total = 0;
  for (let reads = 0; reads < 9; reads += 1) {
    const result = session.readResponse();
    if (
      result === null ||
      !Buffer.isBuffer(result.chunk) ||
      result.chunk.length > NATIVE_ENGINE_MAX_OBSERVATION_RESPONSE_READ_BYTES
    ) {
      throw invalidReplay();
    }
    total += result.chunk.length;
    if (total > MAX_RECORD_BYTES) throw invalidReplay();
    chunks.push(result.chunk);
    if (result.eof) return Buffer.concat(chunks);
  }
  throw invalidReplay();
}

function sameKeys(value, keys) {
  return Object.keys(value).sort().join("\0") === [...keys].sort().join("\0");
}

function require(condition) {
  if (!condition) throw invalidRecord();
}

function invalidRecord() {
  const error = new TypeError("The semantic dependency record is invalid.");
  error.code = "ERR_REPROIT_SEMANTIC_DEPENDENCY";
  return error;
}

function invalidReplay() {
  const error = new Error("The semantic dependency replay is invalid.");
  error.code = "ERR_REPROIT_SEMANTIC_DEPENDENCY";
  return error;
}

File: sdks/python/reproit_sdk/semantic_dependency.py
"""Private semantic dependency bridge to the shared SDK engine."""

from __future__ import annotations

import base64
import inspect
import json
from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass
from typing import Literal

from .native_engine import (
    MAX_OBSERVATION_RESPONSE_READ_BYTES,
    MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES,
)

_ObservationClass = Literal["database", "outbound-http", "queue"]
_Operation = Literal[
    "database-execute",
    "outbound-http-request",
    "queue-acknowledge",
    "queue-publish",
    "queue-receive",
    "queue-reject",
]
_Outcome = Literal["error", "response"]
_ErrorCode = Literal[
    "interrupted",
    "invalid-input",
    "not-found",
    "other",
    "permission-denied",
    "resource-limit",
    "unsupported",
]


class _SemanticDependencyError(RuntimeError):
    pass


@dataclass(frozen=True)
class _DependencyRequest:
    observation_class: _ObservationClass
    operation: _Operation
    protocol: str
    encoding: str
    target: str
    method: str | None
    payload: bytes
    metadata: tuple[tuple[str, bytes], ...] = ()


@dataclass(frozen=True)
class _DependencyResponse:
    outcome: _Outcome
    payload: bytes | None = None
    metadata: tuple[tuple[str, bytes], ...] = ()
    status: str | None = None
    status_code: int | None = None
    error_code: _ErrorCode | None = None
    error_number: int | None = None


def _run_dependency(
    request: _DependencyRequest,
    capture: Callable[[], _DependencyResponse | Awaitable[_DependencyResponse]],
) -> _DependencyResponse | Awaitable[_DependencyResponse]:
    """Drive one private dependency interaction through the shared engine."""
    context = _active_context()
    if context is None:
        return capture()
    try:
        request_input = _request_input(request)
        session = context._open_dependency(request_input)
    except Exception:  # noqa: BLE001 - capture infrastructure must fail open.
        context._abandon()
        return capture()
    if session is None:
        return capture()
    if session.action == "capture":
        return _capture_dependency(session, capture)
    if session.action == "replay":
        return _replay_dependency(session)
    session._abandon()
    return capture()


def _capture_dependency(
    session: object,
    capture: Callable[[], _DependencyResponse | Awaitable[_DependencyResponse]],
) -> _DependencyResponse | Awaitable[_DependencyResponse]:
    try:
        captured = capture()
    except BaseException:
        session._abandon()
        raise
    if inspect.isawaitable(captured):
        return _finish_capture_awaitable(session, captured)
    return _finish_capture(session, captured)


async def _finish_capture_awaitable(
    session: object,
    captured: Awaitable[_DependencyResponse],
) -> _DependencyResponse:
    try:
        response = await captured
    except BaseException:
        session._abandon()
        raise
    return _finish_capture(session, response)


def _finish_capture(
    session: object,
    response: _DependencyResponse,
) -> _DependencyResponse:
    try:
        session._finish(_response_input(response))
    except Exception:  # noqa: BLE001 - capture infrastructure must fail open.
        session._abandon()
    return response


def _replay_dependency(session: object) -> _DependencyResponse:
    try:
        response_record = _read_response(session)
        outcome = session._finish(None)
        if outcome is None:
            raise _invalid_replay()
        response = _response_from_record(response_record)
        if response.outcome != outcome:
            raise _invalid_replay()
        return response
    except Exception as error:
        session._abandon()
        if isinstance(error, _SemanticDependencyError):
            raise
        raise _invalid_replay() from error


def _request_input(request: _DependencyRequest) -> dict[str, object]:
    return {
        "encoding": _bounded_text(request.encoding),
        "metadata": _encode_metadata(request.metadata),
        "method": None if request.method is None else _bounded_text(request.method),
        "observation_class": _bounded_text(request.observation_class),
        "operation": _bounded_text(request.operation),
        "payload": _base64url(_bounded_bytes(request.payload)),
        "protocol": _bounded_text(request.protocol),
        "target": _base64url(_text_bytes(request.target)),
    }


def _response_input(response: _DependencyResponse) -> dict[str, object]:
    payload = response.payload
    return {
        "error_code": response.error_code,
        "error_number": response.error_number,
        "metadata": _encode_metadata(response.metadata),
        "outcome": _bounded_text(response.outcome),
        "payload": None if payload is None else _base64url(_bounded_bytes(payload)),
        "status": None if response.status is None else _bounded_text(response.status),
        "status_code": response.status_code,
    }


def _response_from_record(record: bytes) -> _DependencyResponse:
    try:
        value = json.loads(record)
        if not isinstance(value, dict):
            raise TypeError
        payload = value["payload"]
        return _DependencyResponse(
            outcome=value["outcome"],
            payload=None if payload is None else _decode_base64url(payload),
            metadata=_decode_metadata(value["metadata"]),
            status=value["status"],
            status_code=value["status_code"],
            error_code=value["error_code"],
            error_number=value["error_number"],
        )
    except (
        KeyError,
        TypeError,
        ValueError,
        UnicodeError,
        json.JSONDecodeError,
    ) as error:
        raise _invalid_replay() from error


def _encode_metadata(
    metadata: Sequence[tuple[str, bytes]],
) -> list[dict[str, str]]:
    if not isinstance(metadata, tuple) or len(metadata) > (
        MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES
    ):
        raise _invalid_value()
    encoded = []
    total_bytes = 0
    for field in metadata:
        if not isinstance(field, tuple) or len(field) != 2:
            raise _invalid_value()
        name, value = field
        name_bytes = _text_bytes(name)
        value_bytes = _bounded_bytes(value)
        total_bytes += len(name_bytes) + len(value_bytes)
        if total_bytes > MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES:
            raise _invalid_value()
        encoded.append(
            {"name": _base64url(name_bytes), "value": _base64url(value_bytes)}
        )
    return encoded


def _decode_metadata(value: object) -> tuple[tuple[str, bytes], ...]:
    if not isinstance(value, list) or len(value) > MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES:
        raise _invalid_replay()
    fields = []
    total_bytes = 0
    for field in value:
        if not isinstance(field, Mapping):
            raise _invalid_replay()
        name = _decode_base64url(field.get("name")).decode("utf-8", "strict")
        field_value = _decode_base64url(field.get("value"))
        total_bytes += len(name.encode("utf-8")) + len(field_value)
        if total_bytes > MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES:
            raise _invalid_replay()
        fields.append((name, field_value))
    return tuple(fields)


def _read_response(session: object) -> bytes:
    chunks = []
    total_bytes = 0
    maximum_reads = MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES
    for _ in range(maximum_reads):
        result = session._read_response()
        if result is None:
            break
        chunk, eof = result
        if (
            not isinstance(chunk, bytes)
            or len(chunk) > MAX_OBSERVATION_RESPONSE_READ_BYTES
            or (not chunk and not eof)
        ):
            break
        total_bytes += len(chunk)
        if total_bytes > MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES:
            break
        chunks.append(chunk)
        if eof:
            record = b"".join(chunks)
            if record:
                return record
            break
    raise _invalid_replay()


def _text_bytes(value: object) -> bytes:
    if not isinstance(value, str):
        raise _invalid_value()
    encoded = value.encode("utf-8", "strict")
    if len(encoded) > MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES:
        raise _invalid_value()
    return encoded


def _bounded_text(value: object) -> str:
    _text_bytes(value)
    return value


def _bounded_bytes(value: object) -> bytes:
    if not isinstance(value, bytes) or len(value) > (
        MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES
    ):
        raise _invalid_value()
    return value


def _base64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def _decode_base64url(value: object) -> bytes:
    if not isinstance(value, str) or len(value) > (
        MAX_SEMANTIC_DEPENDENCY_RECORD_BYTES * 2
    ):
        raise _invalid_replay()
    decoded = base64.b64decode(
        value + "=" * (-len(value) % 4),
        altchars=b"-_",
        validate=True,
    )
    if _base64url(decoded) != value:
        raise _invalid_replay()
    return decoded


def _active_context() -> object | None:
    from .engine_operation import _current_operation_context

    return _current_operation_context()


def _invalid_value() -> _SemanticDependencyError:
    return _SemanticDependencyError("The semantic dependency value is invalid.")


def _invalid_replay() -> _SemanticDependencyError:
    return _SemanticDependencyError("The semantic dependency replay is invalid.")
